### PR TITLE
PTFM-18056 java incorrect paginated search behavior

### PR DESCRIPTION
**Pull request by storozhilov-dnanexus**
Hello @psung ,
Here is the test and a respective fix - would be nice if you can look on that.
Thank you very much!

### CODE REVIEW

File: src/java/src/main/java/com/dnanexus/DXSearch.java [hunk 1]
```diff
@@ -900,11 +909,11 @@ public int size() {
         /**
          * Iterator implementation for findDataObjects results.
          */
-        private class ResultIterator
+        public class ResultIterator
```
Comment (psung) [inline src/java/src/main/java/com/dnanexus/DXSearch.java]: Please make this default-visibility and label it "@VisibleForTesting". That indicates it's being made available for tests and doesn't make it part of our public API.

Same goes for pageNo later.

File: src/java/src/test/java/com/dnanexus/DXSearchTest.java [hunk 2]
```diff
@@ -585,6 +586,17 @@ public void testFindDataObjectsWithPaging() {
                 ImmutableList.copyOf(DXSearch.findDataObjects().inProject(testProject)
                         .nameMatchesGlob("foo*").withClassRecord().execute(3));
         Assert.assertEquals(outputRecords, outputRecordStreamWithPaging);
+
+        // Checking an amount of fetched pages
+        DXSearch.FindDataObjectsResult<DXRecord> findResult =
+                DXSearch.findDataObjects().inProject(testProject).nameMatchesGlob("foo*").withClassRecord().execute(3);
+        Iterator<DXRecord> iter = findResult.iterator();
+        List<DXRecord> iterativelyFetchedRecords = Lists.newArrayList();
```
Comment (psung) [inline src/java/src/test/java/com/dnanexus/DXSearchTest.java]: You should be able to automatically populate the list from the iterable:

```
... = Lists.newArrayList(findResult.iterator());
```

Comment (storozhilov-dnanexus) [inline src/java/src/test/java/com/dnanexus/DXSearchTest.java]: Fixed, thank you - very good to know!

File: src/java/src/main/java/com/dnanexus/DXSearch.java (hunk 1 above)
Comment (storozhilov-dnanexus) [inline src/java/src/main/java/com/dnanexus/DXSearch.java]: I tried to make **ResultIterator** private and labeled as *@VisibleForTesting*, but compiler claimed about that:

```
        @VisibleForTesting
        private class ResultIterator
...
[INFO] Compilation failure
/vagrant/src/java/src/test/java/com/dnanexus/DXSearchTest.java:[596,74] error: FindDataObjectsResult.ResultIterator has private access in FindDataObjectsResult
```

If to leave **ResultIterator** public and to make **int pageNo()** method private compiler could not find a respective symbol:

```
        @VisibleForTesting
        private int pageNo() {
            return this.currentPageNo;
        }
...
[INFO] Compilation failure
/vagrant/src/java/src/test/java/com/dnanexus/DXSearchTest.java:[596,95] error: cannot find symbol
```
I tried to label a **PaginatingFindResultIterator** as *@VisibleForTesting*, but it did not help.

Will look to that tomorrow.

Comment (psung) [inline src/java/src/main/java/com/dnanexus/DXSearch.java]: Not private, but default-visibility:

```
@VisibleForTesting
class ResultIterator {

...

@VisibleForTesting
int pageNo() {
```

(The VisibleForTesting doesn't actually mean anything to the compiler-- it's just a convention for humans to understand that the visibility of the method/class was intentionally widened to allow tests to peek at the internals.)

Comment (storozhilov-dnanexus) [inline src/java/src/main/java/com/dnanexus/DXSearch.java]: Fixed. Thank you!

**Pull request merged**